File: sample/src/main/java/com/alamkanak/weekview/sample/apiclient/ApiEvent.java
package com.alamkanak.weekview.sample.apiclient;

import android.graphics.Color;

import com.alamkanak.weekview.WeekViewDisplayable;
import com.alamkanak.weekview.WeekViewEvent;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

import org.jetbrains.annotations.NotNull;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;

/**
 * An event model that was built for automatic serialization from json to object.
 * Created by Raquib-ul-Alam Kanak on 1/3/16.
 * Website: http://alamkanak.github.io
 */
public class ApiEvent implements WeekViewDisplayable<ApiEvent> {

    @Expose @SerializedName("name")
    private String mName;

    @Expose @SerializedName("dayOfMonth")
    private int mDayOfMonth;

    @Expose @SerializedName("startTime")
    private String mStartTime;

    @Expose @SerializedName("endTime")
    private String mEndTime;

    @Expose @SerializedName("color")
    private String mColor;

    public String getName() {
        return mName;
    }

    public void setName(String name) {
        this.mName = name;
    }

    public int getDayOfMonth() {
        return mDayOfMonth;
    }

    public void setDayOfMonth(int dayOfMonth) {
        this.mDayOfMonth = dayOfMonth;
    }

    public String getStartTime() {
        return mStartTime;
    }

    public void setStartTime(String startTime) {
        this.mStartTime = startTime;
    }

    public String getEndTime() {
        return mEndTime;
    }

    public void setEndTime(String endTime) {
        this.mEndTime = endTime;
    }

    public String getColor() {
        return mColor;
    }

    public void setColor(String color) {
        this.mColor = color;
    }

    @NotNull
    public WeekViewEvent<ApiEvent> toWeekViewEvent() {
        SimpleDateFormat sdf = new SimpleDateFormat("HH:mm", Locale.getDefault());
        Date start = new Date();
        Date end = new Date();

        try {
            start = sdf.parse(getStartTime());
        } catch (ParseException e) {
            e.printStackTrace();
        }

        try {
            end = sdf.parse(getEndTime());
        } catch (ParseException e) {
            e.printStackTrace();
        }

        Calendar now = Calendar.getInstance();

        Calendar startTime = (Calendar) now.clone();
        startTime.setTimeInMillis(start.getTime());
        startTime.set(Calendar.YEAR, now.get(Calendar.YEAR));
        startTime.set(Calendar.MONTH, now.get(Calendar.MONTH));
        startTime.set(Calendar.DAY_OF_MONTH, getDayOfMonth());

        Calendar endTime = (Calendar) startTime.clone();
        endTime.setTimeInMillis(end.getTime());
        endTime.set(Calendar.YEAR, startTime.get(Calendar.YEAR));
        endTime.set(Calendar.MONTH, startTime.get(Calendar.MONTH));
        endTime.set(Calendar.DAY_OF_MONTH, startTime.get(Calendar.DAY_OF_MONTH));

        final int color = Color.parseColor(getColor());

        WeekViewEvent.Style style = new WeekViewEvent.Style.Builder()
                .setBackgroundColor(color)
                .build();

        return new WeekViewEvent.Builder<ApiEvent>()
                .setId(0)
                .setTitle(getName())
                .setStartTime(startTime)
                .setEndTime(endTime)
                .setAllDay(false)
                .setStyle(style)
                .setData(this)
                .build();
    }
}

File: sample/src/main/java/com/alamkanak/weekview/sample/data/model/ApiEvent.kt
package com.alamkanak.weekview.sample.data.model

import android.graphics.Color
import com.google.gson.annotations.SerializedName
import java.time.DateTimeException
import java.time.LocalTime
import java.time.YearMonth

interface ApiResult {
    fun toCalendarEntity(yearMonth: YearMonth, index: Int): CalendarEntity?
}

data class ApiEvent(
    @SerializedName("title") val title: String,
    @SerializedName("location") val location: String,
    @SerializedName("day_of_month") val dayOfMonth: Int,
    @SerializedName("start_time") val startTime: String,
    @SerializedName("duration") val duration: Int,
    @SerializedName("color") val color: String,
    @SerializedName("is_canceled") val isCanceled: Boolean,
    @SerializedName("is_all_day") val isAllDay: Boolean
) : ApiResult {

    override fun toCalendarEntity(yearMonth: YearMonth, index: Int): CalendarEntity? {
        return try {
            val startTime = LocalTime.parse(startTime)
            val startDateTime = yearMonth.atDay(dayOfMonth).atTime(startTime)
            val endDateTime = startDateTime.plusMinutes(duration.toLong())
            CalendarEntity.Event(
                id = "100${yearMonth.year}00${yearMonth.monthValue}00$index".toLong(),
                title = title,
                location = location,
                startTime = startDateTime,
                endTime = endDateTime,
                color = Color.parseColor(color),
                isAllDay = isAllDay,
                isCanceled = isCanceled
            )
        } catch (e: DateTimeException) {
            null
        }
    }
}

data class ApiBlockedTime(
    @SerializedName("day_of_month") val dayOfMonth: Int,
    @SerializedName("start_time") val startTime: String,
    @SerializedName("duration") val duration: Int
) : ApiResult {

    override fun toCalendarEntity(yearMonth: YearMonth, index: Int): CalendarEntity? {
        return try {
            val startTime = LocalTime.parse(startTime)
            val startDateTime = yearMonth.atDay(dayOfMonth).atTime(startTime)
            val endDateTime = startDateTime.plusMinutes(duration.toLong())
            CalendarEntity.BlockedTimeSlot(
                id = "200${yearMonth.year}00${yearMonth.monthValue}00$index".toLong(),
                startTime = startDateTime,
                endTime = endDateTime
            )
        } catch (e: DateTimeException) {
            null
        }
    }
}
